File: kagent/go/adk/pkg/embedding/embedding.go
package embedding

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"strings"
	"sync"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/bedrockruntime"
	"github.com/go-logr/logr"
	"github.com/kagent-dev/kagent/go/api/adk"
	"google.golang.org/genai"
)

const (
	// TargetDimension is the required embedding dimension for Kagent memory storage (768)
	TargetDimension = 768
)

// provider is the internal interface for per-provider embedding generation.
type provider interface {
	generate(ctx context.Context, texts []string) ([][]float32, error)
}

// Client generates embeddings using a configured provider.
type Client struct {
	config *adk.EmbeddingConfig
	p      provider
}

// Config for creating an embedding client.
type Config struct {
	EmbeddingConfig *adk.EmbeddingConfig
	HTTPClient      *http.Client
}

// New creates a new embedding client.
func New(cfg Config) (*Client, error) {
	if cfg.EmbeddingConfig == nil {
		return nil, fmt.Errorf("embedding config is required")
	}
	if cfg.EmbeddingConfig.Model == "" {
		return nil, fmt.Errorf("embedding model is required")
	}
	httpClient := cfg.HTTPClient
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	return &Client{
		config: cfg.EmbeddingConfig,
		p:      newProvider(cfg.EmbeddingConfig, httpClient),
	}, nil
}

func newProvider(cfg *adk.EmbeddingConfig, httpClient *http.Client) provider {
	switch cfg.Provider {
	case "azure_openai":
		return &azureOpenAIProvider{config: cfg, httpClient: httpClient}
	case "ollama":
		return &ollamaProvider{config: cfg, httpClient: httpClient}
	case "gemini", "vertex_ai":
		return &geminiProvider{config: cfg}
	case "bedrock":
		return &bedrockProvider{config: cfg}
	default: // "openai", "", and unknown providers
		return &openAIProvider{config: cfg, httpClient: httpClient}
	}
}

// Generate generates embeddings for the given texts.
// Returns a slice of embedding vectors, one per input text.
// Each vector is 768-dimensional (truncated/normalized if needed).
func (c *Client) Generate(ctx context.Context, texts []string) ([][]float32, error) {
	if len(texts) == 0 {
		return nil, fmt.Errorf("no texts provided")
	}
	logr.FromContextOrDiscard(ctx).V(1).Info("Generating embeddings", "count", len(texts), "model", c.config.Model)
	return c.p.generate(ctx, texts)
}

type openAIProvider struct {
	config     *adk.EmbeddingConfig
	httpClient *http.Client
}

func (p *openAIProvider) generate(ctx context.Context, texts []string) ([][]float32, error) {
	log := logr.FromContextOrDiscard(ctx)

	baseURL := p.config.BaseUrl
	if baseURL == "" {
		baseURL = "https://api.openai.com/v1"
	}
	url := fmt.Sprintf("%s/embeddings", baseURL)

	reqBody := map[string]any{
		"input":      texts,
		"model":      p.config.Model,
		"dimensions": TargetDimension,
	}
	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	if apiKey := os.Getenv("OPENAI_API_KEY"); apiKey != "" {
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", apiKey))
	}

	resp, err := p.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("API returned status %d: %s", resp.StatusCode, string(body))
	}

	var result openAIEmbeddingResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	embeddings := make([][]float32, 0, len(result.Data))
	for _, item := range result.Data {
		embedding := item.Embedding
		if len(embedding) > TargetDimension {
			log.V(1).Info("Truncating embedding", "from", len(embedding), "to", TargetDimension)
			embedding = normalizeL2(embedding[:TargetDimension])
		} else if len(embedding) < TargetDimension {
			return nil, fmt.Errorf("embedding dimension %d is less than required %d", len(embedding), TargetDimension)
		}
		embeddings = append(embeddings, embedding)
	}
	log.Info("Successfully generated embeddings", "count", len(embeddings))
	return embeddings, nil
}

type azureOpenAIProvider struct {
	config     *adk.EmbeddingConfig
	httpClient *http.Client
}

func (p *azureOpenAIProvider) generate(ctx context.Context, texts []string) ([][]float32, error) {
	if p.config.BaseUrl == "" {
		return nil, fmt.Errorf("base_url is required for Azure OpenAI")
	}
	url := fmt.Sprintf("%s/embeddings", p.config.BaseUrl)

	reqBody := map[string]any{"input": texts}
	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	if apiKey := os.Getenv("AZURE_OPENAI_API_KEY"); apiKey != "" {
		req.Header.Set("api-key", apiKey)
	}

	resp, err := p.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("API returned status %d: %s", resp.StatusCode, string(body))
	}

	var result openAIEmbeddingResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	embeddings := make([][]float32, 0, len(result.Data))
	for _, item := range result.Data {
		embedding := item.Embedding
		if len(embedding) > TargetDimension {
			embedding = normalizeL2(embedding[:TargetDimension])
		}
		embeddings = append(embeddings, embedding)
	}
	return embeddings, nil
}

type ollamaProvider struct {
	config     *adk.EmbeddingConfig
	httpClient *http.Client
}

func (p *ollamaProvider) generate(ctx context.Context, texts []string) ([][]float32, error) {
	log := logr.FromContextOrDiscard(ctx)

	baseURL := p.config.BaseUrl
	if baseURL == "" {
		baseURL = os.Getenv("OLLAMA_API_BASE")
	}
	if baseURL == "" {
		baseURL = "http://localhost:11434"
	}
	url := fmt.Sprintf("%s/v1/embeddings", strings.TrimSuffix(baseURL, "/"))

	reqBody := map[string]any{
		"input": texts,
		"model": p.config.Model,
	}
	bodyBytes, err := json.Marshal(reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := p.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to make request: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("API returned status %d: %s", resp.StatusCode, string(body))
	}

	var result openAIEmbeddingResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	embeddings := make([][]float32, 0, len(result.Data))
	for _, item := range result.Data {
		embedding := item.Embedding
		if len(embedding) > TargetDimension {
			log.V(1).Info("Truncating embedding", "from", len(embedding), "to", TargetDimension)
			embedding = normalizeL2(embedding[:TargetDimension])
		} else if len(embedding) < TargetDimension {
			return nil, fmt.Errorf("embedding dimension %d is less than required %d", len(embedding), TargetDimension)
		}
		embeddings = append(embeddings, embedding)
	}
	log.Info("Successfully generated embeddings with Ollama", "count", len(embeddings))
	return embeddings, nil
}

type geminiProvider struct {
	config  *adk.EmbeddingConfig
	once    sync.Once
	client  *genai.Client
	initErr error
}

func (p *geminiProvider) generate(ctx context.Context, texts []string) ([][]float32, error) {
	log := logr.FromContextOrDiscard(ctx)

	p.once.Do(func() {
		client, err := genai.NewClient(ctx, &genai.ClientConfig{
			APIKey: os.Getenv("GOOGLE_API_KEY"),
		})
		if err != nil {
			p.initErr = fmt.Errorf("failed to create genai client: %w", err)
			return
		}
		p.client = client
	})
	if p.initErr != nil {
		return nil, p.initErr
	}

	targetDim := int32(TargetDimension)
	embeddings := make([][]float32, len(texts))
	for i, text := range texts {
		result, err := p.client.Models.EmbedContent(ctx, p.config.Model, genai.Text(text), &genai.EmbedContentConfig{
			OutputDimensionality: &targetDim,
		})
		if err != nil {
			return nil, fmt.Errorf("failed to generate embedding for text %d: %w", i, err)
		}
		if len(result.Embeddings) > 0 {
			src := result.Embeddings[0].Values
			emb := make([]float32, len(src))
			for j, v := range src {
				emb[j] = float32(v)
			}
			embeddings[i] = emb
		}
	}
	log.Info("Successfully generated embeddings with Gemini", "count", len(embeddings))
	return embeddings, nil
}

type bedrockProvider struct {
	config  *adk.EmbeddingConfig
	once    sync.Once
	client  *bedrockruntime.Client
	initErr error
}

func (p *bedrockProvider) generate(ctx context.Context, texts []string) ([][]float32, error) {
	log := logr.FromContextOrDiscard(ctx)

	region := os.Getenv("AWS_DEFAULT_REGION")
	if region == "" {
		region = os.Getenv("AWS_REGION")
	}
	if region == "" {
		region = "us-east-1"
	}

	p.once.Do(func() {
		awsCfg, err := awsconfig.LoadDefaultConfig(ctx, awsconfig.WithRegion(region))
		if err != nil {
			p.initErr = fmt.Errorf("failed to load AWS config: %w", err)
			return
		}
		p.client = bedrockruntime.NewFromConfig(awsCfg)
	})
	if p.initErr != nil {
		return nil, p.initErr
	}

	embeddings := make([][]float32, 0, len(texts))
	for i, text := range texts {
		reqBody, err := json.Marshal(map[string]string{"inputText": text})
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request for text %d: %w", i, err)
		}
		output, err := p.client.InvokeModel(ctx, &bedrockruntime.InvokeModelInput{
			ModelId:     aws.String(p.config.Model),
			Body:        reqBody,
			ContentType: aws.String("application/json"),
			Accept:      aws.String("application/json"),
		})
		if err != nil {
			return nil, fmt.Errorf("failed to invoke Bedrock model for text %d: %w", i, err)
		}
		var result bedrockEmbeddingResponse
		if err := json.Unmarshal(output.Body, &result); err != nil {
			return nil, fmt.Errorf("failed to decode Bedrock response for text %d: %w", i, err)
		}
		embedding := result.Embedding
		if len(embedding) > TargetDimension {
			log.V(1).Info("Truncating embedding", "from", len(embedding), "to", TargetDimension)
			embedding = normalizeL2(embedding[:TargetDimension])
		} else if len(embedding) < TargetDimension {
			return nil, fmt.Errorf("embedding dimension %d is less than required %d", len(embedding), TargetDimension)
		}
		embeddings = append(embeddings, embedding)
	}
	log.Info("Successfully generated embeddings with Bedrock", "count", len(embeddings))
	return embeddings, nil
}

type bedrockEmbeddingResponse struct {
	Embedding []float32 `json:"embedding"`
}

type openAIEmbeddingResponse struct {
	Data  []openAIEmbeddingData `json:"data"`
	Model string                `json:"model"`
	Usage openAIUsage           `json:"usage"`
}

type openAIEmbeddingData struct {
	Embedding []float32 `json:"embedding"`
	Index     int       `json:"index"`
}

type openAIUsage struct {
	PromptTokens int `json:"prompt_tokens"`
	TotalTokens  int `json:"total_tokens"`
}

// normalizeL2 normalizes a vector to unit length using L2 norm.
func normalizeL2(vec []float32) []float32 {
	var sum float64
	for _, v := range vec {
		sum += float64(v) * float64(v)
	}
	norm := math.Sqrt(sum)
	if norm == 0 {
		return vec
	}
	normalized := make([]float32, len(vec))
	for i, v := range vec {
		normalized[i] = float32(float64(v) / norm)
	}
	return normalized
}

File: kagent/python/packages/kagent-adk/src/kagent/adk/models/_embedding.py
"""Embedding client for generating vector embeddings using various providers.

This module provides a standalone EmbeddingClient that supports multiple providers:
- openai: OpenAI API embeddings
- azure_openai: Azure OpenAI embeddings
- ollama: Ollama local embeddings
- gemini/vertex_ai: Google Gemini/Vertex AI embeddings
- bedrock: AWS Bedrock Titan Embedding API
"""

import asyncio
import json
import logging
import os
from typing import Any, List, Union

import numpy as np

from kagent.adk.types import EmbeddingConfig

logger = logging.getLogger(__name__)


class KAgentEmbedding:
    """Client for generating embeddings using provider-specific SDKs.

    This client is standalone and has no dependencies on the memory service.
    It supports multiple embedding providers with dimension enforcement and
    L2 normalization.
    """

    # Target dimension for Kagent memory storage (must match go/adk/pkg/embedding/embedding.go)
    TARGET_DIMENSION = 768

    def __init__(self, config: EmbeddingConfig):
        """Initialize EmbeddingClient.

        Args:
            config: Embedding configuration including model, provider, and base_url
        """
        self.config = config

    async def generate(self, texts: Union[str, List[str]]) -> Union[List[float], List[List[float]]]:
        """Generate embedding vector(s) for the given text(s).

        Args:
            texts: Single string or list of strings to embed.

        Returns:
            Single vector (List[float]) if input is string,
            or List of vectors (List[List[float]]) if input is list.
            Returns empty list on failure.
        """
        if not texts:
            return [] if isinstance(texts, list) else []

        is_batch = isinstance(texts, list)
        text_list = texts if is_batch else [texts]

        if not text_list:
            return [] if is_batch else []

        try:
            raw_embeddings = await self._call_provider(text_list)
        except Exception as e:
            logger.error(
                "Error generating embedding with provider=%s model=%s: %s",
                self.config.provider,
                self.config.model,
                e,
            )
            return [] if is_batch else []

        # Enforce dimension consistency and apply L2 normalization
        embeddings = self._process_embeddings(raw_embeddings)

        if is_batch:
            return embeddings
        return embeddings[0] if embeddings else []

    async def _call_provider(self, texts: List[str]) -> List[List[float]]:
        """Dispatch to the correct provider SDK for embedding generation."""
        provider = self.config.provider.lower()

        if provider in ("openai", "azure_openai"):
            return await self._embed_openai(texts)
        if provider == "ollama":
            return await self._embed_ollama(texts)
        if provider in ("vertex_ai", "gemini"):
            return await self._embed_google(texts)
        if provider == "bedrock":
            return await self._embed_bedrock(texts)

        # Unknown provider - try OpenAI-compatible as a fallback
        logger.warning(
            "Unknown embedding provider '%s'; attempting OpenAI-compatible call.",
            provider,
        )
        return await self._embed_openai(texts)

    def _process_embeddings(self, embeddings: List[List[float]]) -> List[List[float]]:
        """Process embeddings to ensure consistent dimensions and L2 normalization.

        Most Matryoshka Representation Learning embedding models produce embeddings
        that still have meaning when truncated to specific sizes:
        https://huggingface.co/blog/matryoshka

        We must ensure embeddings have consistent dimensions for the vector storage backend.
        """
        processed = []

        for embedding in embeddings:
            dim = len(embedding)
            processed_embedding = embedding

            if dim > self.TARGET_DIMENSION:
                # Truncate to target dimension
                processed_embedding = embedding[: self.TARGET_DIMENSION]
                # Re-normalize after truncation
                processed_embedding = self._normalize_l2(processed_embedding).tolist()
            elif dim < self.TARGET_DIMENSION:
                logger.error(
                    "Embedding dimension %d is smaller than required %d; rejecting embeddings batch",
                    dim,
                    self.TARGET_DIMENSION,
                )
                return []

            processed.append(processed_embedding)

        return processed

    def _normalize_l2(self, x: Union[List[float], np.ndarray]) -> np.ndarray:
        """Apply L2 normalization to a vector or array of vectors."""
        x = np.array(x)
        if x.ndim == 1:
            norm = np.linalg.norm(x)
            if norm == 0:
                return x
            return x / norm
        else:
            norm = np.linalg.norm(x, 2, axis=1, keepdims=True)
            return np.where(norm == 0, x, x / norm)

    async def _embed_openai(self, texts: List[str]) -> List[List[float]]:
        """Embed using the OpenAI or Azure OpenAI SDK."""
        provider = self.config.provider.lower()

        if provider == "azure_openai":
            from openai import AsyncAzureOpenAI

            api_version = os.environ.get("OPENAI_API_VERSION", "2024-02-15-preview")
            api_base = self.config.base_url or os.environ.get("AZURE_OPENAI_ENDPOINT")
            if not api_base:
                raise ValueError("Azure OpenAI endpoint must be set via base_url or AZURE_OPENAI_ENDPOINT env var")
            client = AsyncAzureOpenAI(api_version=api_version, azure_endpoint=api_base)
        else:
            from openai import AsyncOpenAI

            client = AsyncOpenAI(base_url=self.config.base_url or None)

        response = await client.embeddings.create(
            model=self.config.model,
            input=texts,
            dimensions=self.TARGET_DIMENSION,
        )
        return [item.embedding for item in response.data]

    async def _embed_ollama(self, texts: List[str]) -> List[List[float]]:
        """Embed using the Ollama SDK."""
        import ollama

        host = self.config.base_url or os.environ.get("OLLAMA_API_BASE", "http://localhost:11434")
        client = ollama.AsyncClient(host=host)
        result = await client.embed(model=self.config.model, input=texts)
        # Ollama returns embeddings as a list of lists
        embeddings = result.embeddings
        if embeddings and not isinstance(embeddings[0], list):
            # Single embedding case
            return [embeddings]
        return list(embeddings)

    async def _embed_google(self, texts: List[str]) -> List[List[float]]:
        """Embed using google-genai (Gemini or Vertex AI)."""
        from google import genai
        from google.genai import types as genai_types

        if self.config.provider.lower() == "vertex_ai":
            client = genai.Client(vertexai=True)
        else:
            client = genai.Client()

        # Use asyncio.to_thread since genai may not have async methods
        response = await asyncio.to_thread(
            client.models.embed_content,
            model=self.config.model,
            contents=texts,
            config=genai_types.EmbedContentConfig(output_dimensionality=self.TARGET_DIMENSION),
        )
        return [list(emb.values) for emb in response.embeddings]

    async def _embed_bedrock(
        self,
        texts: List[str],
    ) -> List[List[float]]:
        """Embed using the AWS Bedrock Titan Embedding API via boto3.

        Uses the same credential chain (env vars, IRSA, instance profile) as
        KAgentBedrockLlm.  Each text is embedded individually because the
        Titan Embedding API accepts a single ``inputText`` per invocation.
        """
        import boto3

        region = os.environ.get("AWS_DEFAULT_REGION") or os.environ.get("AWS_REGION") or "us-east-1"
        client = boto3.client("bedrock-runtime", region_name=region)

        async def _invoke_single(text: str) -> List[float]:
            body = json.dumps({"inputText": text})
            response = await asyncio.to_thread(
                client.invoke_model,
                modelId=self.config.model,
                body=body,
                contentType="application/json",
                accept="application/json",
            )
            result = json.loads(response["body"].read())
            return result["embedding"]

        embeddings = await asyncio.gather(*[_invoke_single(t) for t in texts])
        return list(embeddings)
